### ISSUE TITLE
RFC: Integrating Debugging Data Collection into VERL

### ISSUE THREAD

**Issue by Tjh-UKN** [verl-project/verl#5808]
### Feature request

This RFC proposes a configurable, opt-in, low-intrusion precision debugging layer for VERL. When enabled through external configuration, VERL will collect intermediate numerical data at selected execution stages and persist it for offline analysis. Typical use cases include debugging training-serving mismatches, NaN/Inf propagation, overflow, dtype inconsistencies, and operator-level numerical differences.

When disabled, the feature must introduce no dependency requirement, no runtime overhead, and no behavior change.

The design is backend-pluggable. The initial target backend is `msprobe` / MindStudio Probe PrecisionDebugger, but the VERL-side integration should remain backend-agnostic.

### Motivation

Debugging numerical issues in large-scale RLHF and post-training pipelines is expensive and slow. Problems often appear only in distributed or mixed-precision runs, and existing logs are usually insufficient to localize the failure. Developers need a way to inspect intermediate tensors, statistics, and gradients at critical stages without permanently instrumenting the training stack.

`msprobe` already provides a practical precision debugging capability across multiple backends, including vLLM (eager), FSDP, and Megatron. It supports:

- Collecting intermediate model data, such as per-layer outputs and log-probability differences.
- Tensor-level collection for detailed debugging.
- Lightweight statistics-only collection for faster runs and smaller dumps.
- Dynamic `start` / `stop` control around selected stages.
- Post-processing workflows such as visualization and comparison.

This RFC proposes a VERL-native integration layer so that these capabilities can be enabled through standard VERL configuration and stage boundaries.

## Goals

- Support intermediate data collection for the following stages, each independently configurable:
  - Rollout stage, centered on `generate_sequences`
  - Training forward stage, including loss/logprob/value-related forward computation
  - Reference model stage, centered on `compute_ref_log_prob`
  - Training backward stage, centered on `backward`
  - Parameter update stage, centered on `optimizer.step` and before/after update state
- Keep the feature disabled by default.
- Guarantee that the disabled path has no dependency on the backend package and no meaningful runtime overhead.
- Drive behavior entirely through external configuration, including:
  - Sampling by step, rank, and stage
  - White/blacklist-based module or operator selection
  - Structured output directories and metadata such as `global_step`, `uid`, `rank`, and `stage`
- Expose the integration as a pluggable backend interface rather than hard-coding `msprobe` into VERL.

## Non-Goals

- Replacing existing profilers or performance tracing tools.
- Standardizing the internal configuration schema of all third-party debuggers.
- Guaranteeing identical instrumentation granularity across all execution backends in the first iteration.
- Solving all service-side rollout instrumentation gaps immediately for every runtime.

## Design Principles

- Minimal intrusion: integrate at stable stage boundaries rather than scattering hooks throughout the codebase.
- Backend abstraction: VERL owns stage lifecycle and metadata; the debugger backend owns collection details.
- Zero-cost disabled path: no backend import, initialization, or conditional work on the hot path unless explicitly enabled.
- Incremental rollout: land the abstraction first, then the backend, then stage-specific integrations.

### 1. Add a generic backend interface

Introduce a lightweight precision debugger backend interface:

```python
class PrecisionDebuggerBackend:
    def enabled(self) -> bool: ...

    @contextmanager
    def stage(self, name: str, *, model=None, meta: dict | None = None): ...

    def step(self, *, meta: dict | None = None): ...
```

Planned implementations:

- `NoopPrecisionDebugger`: default implementation; all methods are no-ops.
- `MsprobePrecisionDebugger`: wraps `msprobe` lifecycle APIs such as `start`, `stop`, and per-step progression.

The VERL codebase should depend only on this interface.

### 2. Keep `msprobe` as an optional dependency

The `msprobe` package should be imported only when the debugger is enabled and the selected backend is `msprobe`.

Recommended packaging model:

- Default VERL install does not require `mindstudio-probe`
- Optional install path such as `pip install verl[msprobe]`

This avoids polluting the default dependency graph and ensures that users who do not need the feature are unaffected.

### 3. Instrument stage boundaries at the worker layer

The preferred integration strategy is to wrap stage entry points at the worker layer. This matches VERL's execution model, where the driver in `verl/trainer/ppo/ray_trainer.py` dispatches stage execution through worker groups.

Worker-layer integration provides a single place to cover multiple implementations, including FSDP, Megatron, and engine workers.

Initial integration points:

- Rollout:
  - `verl/workers/{megatron_workers,fsdp_workers}.py:generate_sequences`
- Reference model:
  - `verl/workers/{megatron_workers,engine_workers,fsdp_workers}.py:compute_ref_log_prob`
- Actor update:
  - `verl/workers/{megatron_workers,engine_workers,fsdp_workers}.py:update_actor`

VERL already uses stable stage naming in the trainer, for example `gen`, `ref`, and `update_actor` timing. Reusing consistent stage names reduces cognitive overhead for users.

### 4. Use coarse and fine-grained stages

Two levels of instrumentation are proposed:

- Coarse-grained worker stages:
  - `rollout`
  - `ref_model`
  - `update_actor`
- Fine-grained training sub-stages inside actor implementations:
  - `train_forward`
  - `train_backward`
  - `update_step`

For FSDP actors, these boundaries naturally align with `forward -> loss -> backward -> optimizer_step`.

For Megatron actors, the first version may wrap a larger update boundary such as `update_policy` or `forward_backward_batch`, then refine later if needed.

### VERL configuration

Add a top-level configuration block, parallel to existing profiler-related settings:

```yaml
precision_debugger:
  enable: false

  backend: msprobe  # or null

  # VERL-side gates to limit collection scope
  stages: ["rollout", "train_forward", "ref_model", "train_backward", "update_actor"]
  ranks: null       # null = all, or e.g. [0, 1]
  steps: null       # null = all, or e.g. [100, 200]

  # Backend-specific config file path, interpreted by the backend
  config_path: null

  # Root dump directory
  dump_root: "outputs/precision_debug"

  # If backend import/init fails:
  # true  -> continue training with a warning
  # false -> fail fast
  fail_open: true
```

Notes:

- `enable: false` selects the no-op path.
- `stages`, `ranks`, and `steps` act as VERL-side gates independent of backend-specific filtering.
- `config_path` and `dump_root` are passed to the selected backend.

### Backend configuration

`msprobe` should continue using its native configuration format for collection details such as:

- Which modules or operators to trace
- Whether to collect full tensors or statistics only
- Dump formatting and backend-specific collection policies

VERL should not duplicate that schema. Instead:

- VERL is responsible for calling `start` / `stop` / `step` at the correct stage boundaries.
- VERL passes stage metadata such as `stage`, `global_step`, `rank`, and `uid`.
- The backend is responsible for interpreting its own configuration and organizing collected artifacts.

## Output Layout and Metadata

To make the collected data immediately usable, the backend should follow a predictable directory structure:

```text
outputs/precision_debug/
  exp=<experiment_name>/
    stage=<rollout|ref_model|train_forward|train_backward|update_actor>/
      step=<global_step>/
        rank=<global_rank>/
          ... backend dumps ...
        meta.json
```

Suggested metadata fields:

- `stage`
- `global_step`
- `rank`
- `world_size`
- Optional parallelism coordinates such as `dp_rank`, `tp_rank`, and `pp_rank`
- Batch `uid` values, if available
- Dtype / AMP summary, such as `bf16`, `fp16`, `fp8`, and loss scaling information
- Config summary and commit SHA, where practical

### Performance requirements

The feature must remain operationally safe to enable in production debugging scenarios:

- Disabled by default
- Scope-limited by `stages`, `steps`, and `ranks`
- Compatible with low-volume sampling, such as:
  - only selected UIDs per step
  - a fixed sampling ratio
  - only designated collector ranks, for example one rank per DP group

### Reliability requirements

- All stage wrappers must use `try/finally` so backend `stop()` calls are not skipped on exceptions.
- If `fail_open=true`, backend import or initialization errors should downgrade to a warning and training should continue.
- Collection should default to tensors and numerical statistics, not raw prompt text.
- If text dumping is supported by a backend, it should be explicitly documented as debug-only and user-responsible from a data governance perspective.

## Current Status

Some parts of the capability are already connected in internal experiments:

- `generate_sequences` collection
- `update_actor` collection
- `compute_log_prob` collection
- `compute_ref_log_prob` collection

The target dump structure has also been validated in prototype form.

## Additional Considerations for VERL 0.7+

VERL 0.7 and later introduced a more service-oriented execution model. As a result, directly inserting hooks into some actor-side code paths is no longer sufficient for complete rollout-stage collection.

There are two practical implications:

1. Current PrecisionDebugger support is focused on `vllm-ascend` via service-side configuration, while `GPUModelRunner` coverage remains incomplete.
2. A more uniform solution may require patching utilities under `verl/workers/rollout/vllm_rollout/utils.py` so rollout-side instrumentation can be applied consistently.

This patch-based utility approach must still preserve the core requirement: when disabled, it must impose effectively zero performance overhead.

### Server-side global step synchronization

To place rollout dumps under the correct `global_step` directory, the rollout service side also needs access to the current global step.

One practical approach is to synchronize a server-side global step counter whenever `update_weights` is called, potentially via a lightweight patch. This ensures rollout-generated debug data and trainer-side steps remain aligned in the dump layout.

## Rollout Plan

1. Infrastructure PR
   Add the `PrecisionDebuggerBackend` abstraction, no-op implementation, and configuration parsing.
2. Backend PR
   Add `MsprobePrecisionDebugger` with lazy import, lifecycle control, and dump directory management.
3. Integration PRs
   Add instrumentation to:
   - `generate_sequences`
   - `compute_ref_log_prob`
   - `update_actor`
   Start with coarse wrapping for actor updates, then refine into `train_forward`, `train_backward`, and `update_step` where practical.
4. Documentation PR
   Add a minimal example configuration, operational guidance, and common troubleshooting notes such as:
   - collecting only a single step
   - collecting only selected ranks
   - limiting dump size

## Open Questions

- Should VERL define a standard stage taxonomy that all backends must follow exactly, or allow backend-specific extensions?
- What is the minimum metadata contract required for cross-run comparison tools?
- Should VERL expose first-class sampling by UID, or leave that entirely to backend configuration?
- How should service-side rollout instrumentation evolve for non-Ascend runtimes?

### Directly integrate `msprobe` APIs throughout VERL

Rejected because it tightly couples VERL to one debugger backend and makes future backends harder to support.

### Instrument only the trainer layer

Rejected because worker-layer execution boundaries are a better match for how VERL dispatches stage work across FSDP, Megatron, and engine worker implementations.

### Always-on instrumentation hooks

Rejected because the disabled path must remain effectively free in both dependencies and runtime cost.

## References

- MindStudio Probe / `mindstudio-probe`: https://gitcode.com/Ascend/msprobe

**Comment by Kirrito-k423** [comment]
Thank you for this RFC on precision debugging integration.

**Assessment of the proposal:**

The opt-in, low-intrusion design principle is crucial — ensuring zero runtime overhead when disabled makes this practical for production training runs while enabling detailed debugging when needed.

**Key considerations:**

1. **Backend-agnostic architecture**: The pluggable backend design (initially `msprobe`) provides flexibility for future integrations with other debugging tools (e.g., PyTorch's built-in profiling, NVIDIA Nsight).

2. **Collection scope**: Targeting intermediate numerical data at selected execution stages allows focused debugging without overwhelming storage. Suggested stages:
   - Rollout generation (logits, sampling decisions)
   - Reward computation (intermediate reward signals)
   - PPO updates (policy ratios, value predictions)
   - Gradient computation (grad norms, overflow detection)

3. **NaN/Inf propagation tracking**: This is particularly valuable for RLHF debugging where reward spikes or policy divergence can cascade into numerical instability.

4. **Training-serving mismatch detection**: Comparing rollout engine outputs vs. training engine expectations could catch weight sync issues early.

**Implementation notes:**

- External configuration trigger (e.g., `--debug_precision=true`) ensures easy activation
- Backend detection via conditional import prevents dependency bloat
- Storage strategy should consider distributed training context (per-rank vs. aggregated)

This would be a valuable diagnostic capability for complex RL training scenarios. Looking forward to seeing the implementation design.
